File: patterns/_tests/StructuralPatterns/Facade/typescript/integration.test.ts
// ============================================================
//  Integration Tests — Facade: Third Party API Hub
//
//  What we test here:
//    · Full makePurchase() flow produces correct end-to-end output
//    · Auth result (email) feeds into Payment, Payment ref feeds into Messaging
//    · Each facade routes exclusively to its own TP base URL
//    · Credentials are masked with ••• — never raw values in output
//    · Swapping a facade subclass leaves other facades untouched
//
//  Run: npx mocha --require ts-node/register integration.test.ts
// ============================================================

import { describe, it, before } from 'mocha';
import assert from 'assert';
import { captureOutput, findMissing } from '../../../_helpers/typescript/helpers';
import {
  PaymentFacade, AuthFacade, MessagingFacade, PurchaseService,
} from '../../../../StructuralPatterns/Facade/typescript/typescript';

// ── Full purchase flow ────────────────────────────────────────
describe('PurchaseService — full makePurchase() flow', () => {
  let lines: string[];

  before(() => {
    lines = captureOutput(() =>
      new PurchaseService().makePurchase('tok_user_jane', 49.99, '+15550142')
    );
  });

  it('all three confirmation lines appear', () => {
    assert.strictEqual(
      findMissing(lines, ['✓ User:', '✓ Charged', '✓ Message queued']),
      null
    );
  });

  it('auth email propagates into consumer output', () => {
    assert.ok(lines.some(l => l.includes('jane.doe@example.com')));
  });

  it('payment transaction id propagates into consumer output', () => {
    assert.ok(lines.some(l => l.includes('txn_8821')));
  });

  it('message id propagates into consumer output', () => {
    assert.ok(lines.some(l => l.includes('msg_3301')));
  });

  it('transaction id appears in the messaging confirmation (data flows between facades)', () => {
    const msgLine = lines.find(l => l.includes('✓ Message queued'));
    // The message body includes the transaction ref — confirms auth→payment→messaging data flow
    const confirmLine = lines.find(l => l.includes('Payment confirmed') || l.includes('txn_8821'));
    assert.ok(msgLine || confirmLine);
  });
});

// ── Loose coupling — base URLs stay inside facades ───────────
describe('Facade — loose coupling (TP details never leak)', () => {
  it('each facade routes exclusively to its own base URL', () => {
    const paymentLines = captureOutput(() => new PaymentFacade().charge('usr_1', 100));
    const authLines    = captureOutput(() => new AuthFacade().validateToken('tok_1'));
    const msgLines     = captureOutput(() => new MessagingFacade().send('+1', 'hi'));

    assert.ok(paymentLines.some(l => l.includes('paymentprocessortp.com')));
    assert.ok(authLines.some(l => l.includes('authprovidertp.com')));
    assert.ok(msgLines.some(l => l.includes('messagingtp.com')));
  });

  it('no cross-contamination between facade URLs', () => {
    const paymentLines = captureOutput(() => new PaymentFacade().charge('usr_1', 100)).join('\n');
    const authLines    = captureOutput(() => new AuthFacade().validateToken('tok_1')).join('\n');

    assert.ok(!paymentLines.includes('authprovidertp.com'));
    assert.ok(!paymentLines.includes('messagingtp.com'));
    assert.ok(!authLines.includes('paymentprocessortp.com'));
    assert.ok(!authLines.includes('messagingtp.com'));
  });

  it('headers are masked — raw credential values never appear in output', () => {
    const out = captureOutput(() =>
      new PurchaseService().makePurchase('tok_test', 10, '+1')
    ).join('\n');
    assert.ok(!out.includes('sk_live_demo'));
    assert.ok(!out.includes('auth_secret_demo'));
    assert.ok(!out.includes('SID_xyz'));
    assert.ok(!out.includes('tok_xyz'));
    // Headers ARE logged but values are masked with •••
    assert.ok(out.includes('•••'));
  });
});

// ── Single point of change ───────────────────────────────────
describe('Facade — single point of change', () => {
  it('swapping PaymentFacade subclass does not affect AuthFacade', () => {
    class FreePaymentFacade extends PaymentFacade {
      override charge(_userId: string, _amount: number) {
        return { transaction_id: 'free_txn', status: 'success' };
      }
    }
    // Auth unaffected
    const authResult = new AuthFacade().validateToken('tok_1');
    assert.strictEqual(authResult['user_id'], 'usr_4421');

    // Swapped payment facade returns custom result
    const txResult = new FreePaymentFacade().charge('usr_1', 0);
    assert.strictEqual(txResult['transaction_id'], 'free_txn');
  });

  it('swapping MessagingFacade subclass does not affect PaymentFacade', () => {
    class SilentMessaging extends MessagingFacade {
      override send(_to: string, _msg: string) {
        return { message_id: 'silent', status: 'dropped' };
      }
    }
    // Payment unaffected
    const txResult = new PaymentFacade().charge('usr_1', 50);
    assert.strictEqual(txResult['transaction_id'], 'txn_8821');

    // Swapped messaging returns custom result
    const msgResult = new SilentMessaging().send('+1', 'test');
    assert.strictEqual(msgResult['status'], 'dropped');
  });
});

File: patterns/_tests/StructuralPatterns/Facade/ruby/integration.test.rb
# ============================================================
#  Integration Tests — Facade: Third Party API Hub
#
#  What we test here:
#    · Full make_purchase() flow produces correct end-to-end output
#    · Auth result feeds into Payment, Payment ref feeds into Messaging
#    · Each facade routes exclusively to its own TP base URL
#    · Credentials masked with ••• — no raw values in output
#    · Swapping a facade subclass leaves other facades untouched
#
#  Run: rspec integration.test.rb
# ============================================================

require 'rspec'
require 'stringio'
require_relative '../../../../StructuralPatterns/Facade/ruby/ruby'

def capture_output(&block)
  old = $stdout
  $stdout = StringIO.new
  block.call
  $stdout.string.lines.map(&:chomp)
ensure
  $stdout = old
end

# ── Full purchase flow ────────────────────────────────────────
RSpec.describe 'PurchaseService — full make_purchase() flow' do
  let(:lines) do
    capture_output { PurchaseService.new.make_purchase('tok_user_jane', 49.99, '+15550142') }
  end

  it 'all three confirmation lines appear' do
    out = lines.join("\n")
    expect(out).to include('✓ User:')
    expect(out).to include('✓ Charged')
    expect(out).to include('✓ Message queued')
  end

  it 'auth email propagates into output' do
    expect(lines).to include(a_string_including('jane.doe@example.com'))
  end

  it 'transaction id propagates into output' do
    expect(lines).to include(a_string_including('txn_8821'))
  end

  it 'message id propagates into output' do
    expect(lines).to include(a_string_including('msg_3301'))
  end
end

# ── Loose coupling ────────────────────────────────────────────
RSpec.describe 'Facade — loose coupling' do
  it 'each facade routes exclusively to its own base URL' do
    payment_lines = capture_output { PaymentFacade.new.charge('usr_1', 100) }
    auth_lines    = capture_output { AuthFacade.new.validate_token('tok_1') }
    msg_lines     = capture_output { MessagingFacade.new.send('+1', 'hi') }

    expect(payment_lines).to include(a_string_including('paymentprocessortp.com'))
    expect(auth_lines).to    include(a_string_including('authprovidertp.com'))
    expect(msg_lines).to     include(a_string_including('messagingtp.com'))
  end

  it 'no cross-contamination between facade URLs' do
    payment_out = capture_output { PaymentFacade.new.charge('usr_1', 100) }.join("\n")
    auth_out    = capture_output { AuthFacade.new.validate_token('tok_1') }.join("\n")

    expect(payment_out).not_to include('authprovidertp.com')
    expect(payment_out).not_to include('messagingtp.com')
    expect(auth_out).not_to    include('paymentprocessortp.com')
  end

  it 'headers are masked — raw credentials never appear' do
    out = capture_output { PurchaseService.new.make_purchase('tok', 10, '+1') }.join("\n")
    expect(out).not_to include('sk_live_demo')
    expect(out).not_to include('auth_secret_demo')
    expect(out).not_to include('SID_xyz')
    expect(out).to include('•••')
  end
end

# ── Single point of change ───────────────────────────────────
RSpec.describe 'Facade — single point of change' do
  it 'swapping PaymentFacade subclass does not affect AuthFacade' do
    free_payment = Class.new(PaymentFacade) do
      def charge(user_id, amount)
        { transaction_id: 'free_txn', status: 'success' }
      end
    end

    auth_result = AuthFacade.new.validate_token('tok_1')
    expect(auth_result[:user_id]).to eq('usr_4421')

    tx = free_payment.new.charge('usr_1', 0)
    expect(tx[:transaction_id]).to eq('free_txn')
  end

  it 'swapping MessagingFacade does not affect PaymentFacade' do
    silent_msg = Class.new(MessagingFacade) do
      def send(to, message)
        { message_id: 'silent', status: 'dropped' }
      end
    end

    tx = PaymentFacade.new.charge('usr_1', 50)
    expect(tx[:transaction_id]).to eq('txn_8821')

    msg = silent_msg.new.send('+1', 'test')
    expect(msg[:status]).to eq('dropped')
  end
end
